### ISSUE TITLE
Organization and access to central program repository (delete, search, duplicate, print)

### ISSUE THREAD

**Issue by BLFiedler** [phetsims/paper-land#8]
We'll want to be able to do more to the database than is currently possible in Paper Programs.

@jessegreenberg and @jbphet successfully connected a hosted database to the tool and are investigating how to expose and give more access to the local spaces (so we could have more than one "repo" of scripts with different purposes or for different audiences)

We will definitely want convenient ways to Delete, Search for, duplicate, and some organizational method for versioning scripts.

We might consider creating a "Template" category or space where the most general examples go, with any necessary subcategories (model (maybe subcategory again: physics models, etc) , speech display, UI component).

Could the Templates go in their own space? Or just have their own label?

Can we restrict editing access to Templates, but leave duplicating and editing those open without restriction?

@jessegreenberg noted that when you are using a shared DB, the tool already locks out users who are editing a script. Two connections cannot edit the same script at the same time.

**Comment by BLFiedler** [comment]
Tagging issue #5 as related.

**Comment by BLFiedler** [comment]
We considered an organization based on "play spaces" (localspace) and searchable keyword tags for programs.

![image](https://user-images.githubusercontent.com/47331395/217349602-ddf7ee0c-f5fb-458a-a417-b13ce33cbb8d.png)

We should have one playspace that contains all programs, one "vetted" space with templates, and then infinite others.

We'll need a way for spaces to reach across each other, e.g., when in summer-codesign-2023, we should be able to pull scripts from approved-templates. We will need another issue for this.

**Comment by BLFiedler** [comment]
Currently this needs Design time to determine the naming and convention for Playspaces we intend to use. We also need to determine what functionality for manipulating Playspaces should be gated by some form of authentication. This is more important for remote DB than local.

**Comment by BLFiedler** [comment]
I made some notes in our meeting doc about the future organization of the database and interface needs, both short and possible long term. Duplicating below with some edits made during a meeting with JB and JG.

### Database Organization - Needed Features ($ high priority)
Programs organized by playspace, filterable by keyword

#### Playspaces: 
$ - [ ] $ “Highlighted” playspaces: Stay on top of playspace list, with divider in dropdown box
templates :  complete, functional examples
Other playspaces: sort by Last Modified? Alphabetical right now. Defer this for now.

#### Keywords: https://github.com/phetsims/papyrus/issues/36
$ User-generated (need to watch for typos)
When we have generated enough keywords, we can group and theme as we see appropriate. Perhaps then we provide a predefined list that is recognized. Difficult to anticipate the breadth of needed categories right now.

#### Authentication:
Remote Database (current behavior, no actions needed)
- Creating a playspace does not require a password, assuming you have access to the DB (which already requires authentication)
- Deleting a playspace can only happen by the DB admin

Local Database (needs local database behavior restored) (low priority) #30 
- Legacy behavior, behaves similarly to above, but they’ll be administering their own locally hosted SQL database (Postgres supported)

#### Program Creation:
Duplication from an existing Program (e.g., copy AltitudeSlider from templates)
- [ ] $ Replace current Create New Program behavior from Hello World default template
  - JB wants to investigate Material UI (React library, material design from Google, offers dialog box). 
    - Create from Template
      - Can we use this? With class based React?
 


Assigning @jbphet for the two task items above for anchoring some playspaces to the top of the playspace list and creating a dialog for creating a new program from a template.

**Comment by jbphet** [comment]
@BLFiedler - @jessegreenberg and I have added the dialog and associated functionality for creating new programs based on existing ones.  The ability to search and filter programs was also added under #36.  Do you still feel like adding the feature to anchor some playspaces to the top is still desirable and high priority?

**Comment by BLFiedler** [comment]
I mentioned in the last design meeting that we will want to be pretty lax with creating playspaces. I anticipate the number of them will grow quite a bit for every "set of examples" we want to create. We don't have a way to search "across" playspaces at the moment, at least not very seamlessly. So navigating through that dropdown is pretty frequent.

Anchoring some to the top definitely feels desirable as a right-now solution, but maybe more seamless search and preview across playspaces is what we actually want?

That said, I don't think it's as high priority as other issues that have come up. Let's plan to discuss. Then we can either continue this issue or close and break off other issues.
